scripts/properly_populate_transaction_fees.js

```diff
@@ -0,0 +1,281 @@
+/**
+ * These are some recommendations for using the script in
+ * *development* mode. We currently have ~70000 transactions in our
+ * database, running it with a large batch to process all the
+ * transactions at once is the fastest way to run it and get the
+ * output, however it also exhausts node's memory limits with batches
+ * larger than ~50000. Use --max-old-space=8192 to do it all in one
+ * batch and iterate faster.
+ *
+ * For production usage, It might be a good idea to break it down in
+ * multiple batches to leave some room for the database to process
+ * operations from the web clients. I've been running 7 batches of
+ * 10000 and didn't take more than ~20s in my machine (i7~2.2GHz),
+ * which is just a little bit slower than running it all in one batch.
+ */
+import { ArgumentParser } from 'argparse';
+import models, { sequelize } from '../server/models';
+import * as transactionsLib from '../server/lib/transactions';
+import * as paymentsLib from '../server/lib/payments';
+import { OC_FEE_PERCENT } from '../server/constants/transactions';
+
+class Migration {
+  constructor(options) {
+    this.options = options;
+    this.offset = 0;
+    this.migrated = 0;
+  }
+
+  /** Retrieve the total number of valid transactions */
+  countValidTransactions = async () => {
+    return models.Transaction.count({ where: { deletedAt: null } });
+  }
+
+  /** Retrieve a batch of valid transactions */
+  retrieveValidTransactions = async () => {
+    const transactions = await models.Transaction.findAll({
+      where: { deletedAt: null },
+      order: ['TransactionGroup'],
+      limit: this.options.batchSize,
+      offset: this.offset,
+      include: [{ model: models.Collective, as: 'collective' }]
+    });
+    this.offset += transactions.length;
+    return transactions;
+  }
+
+  /** Convert `value` to negative if it's possitive */
+  toNegative = (value) => value > 0 ? -value : value;
+
+  /** Ensure that `tr` has the `hostCurrencyFxRate` field filled in */
+  ensureHostCurrencyFxRate = (tr) => {
+    if (tr.amount === tr.amountInHostCurrency && !tr.hostCurrencyFxRate)
+      tr.hostCurrencyFxRate = 1;
```

Comment:
This is a minor thing but I feel like, in general, it's better to make these type of functions as `get` functions, instead of modifying the underlying model. If it's `getHostCurrencyFXRate`, you don't need to think about whether model was saved or not. Otherwise, whoever calls this function will need to remember that it modifies the passed in object and they need to save it.

Reply:
Yeah, in that case what I usually like is to create two layers.

 1. Function for the calculation of the new value (the `get` function you mentioned);
 2. Function that actually performs the mutation in the field (the one that I did);

I started from the second layer because I really wanted to test the actual side effect since these objects are going to be instances of database rows and whatever change to them will actually reflect in the database.

Although this is a script, I actually ended up testing most of the delicate functions because I was thinking about keeping that script around since it became sort of a cleanup tool. But I totally agree with you that we should also have the `get*` functions within a library (thinking of `lib/transactions`). I'll do that in small chunks for things that I need to reuse during the changes in the application code.